File: LeetCode/[M] - 649. Dota2 Senate/solution.py
from collections import deque


class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        senateL = len(senate)
        queueR, queueD = deque(), deque()
        for i in range(len(senate)):
            if senate[i] == "R":
                queueR.append(i)
            else:
                queueD.append(i)
        cntR, cntD = len(queueR), len(queueD)
        while cntR != 0 and cntD != 0:
            ri, di = queueR.popleft(), queueD.popleft()
            if ri < di:
                queueR.append(ri + senateL)
                cntD -= 1
            else:
                queueD.append(di + senateL)
                cntR -= 1
        return "Radiant" if cntD == 0 else "Dire"

File: LeetCode/[M] - 649. Dota2 Senate/solution.go
package main

func predictPartyVictory(senate string) string {
	senateL := len(senate)
	queueR, queueD := make([]int, 0), make([]int, 0)
	for i := 0; i < senateL; i++ {
		if senate[i] == 'R' {
			queueR = append(queueR, i)
		} else {
			queueD = append(queueD, i)
		}
	}
	cntR, cntD := len(queueR), len(queueD)
	for cntR != 0 && cntD != 0 {
		ri, di := queueR[0], queueD[0]
		if ri < di {
			queueR = append(queueR, ri+senateL)
			cntD--
		} else {
			queueD = append(queueD, di+senateL)
			cntR--
		}
		queueR, queueD = queueR[1:], queueD[1:]
	}
	if cntR == 0 {
		return "Dire"
	}
	return "Radiant"
}
